util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+
+		FT_DEBUG("CLIENT: connecting to <%s>\n", ct->opts.dst_addr);
+		ret = connect(ct->ctrl_connfd, (struct sockaddr *)&ct->ctrl_addr, sizeof(ct->ctrl_addr));
+		if (ret == -1) {
+			err = errno;
+			FT_PRINTERR("connect", err);
+			return ret;
+		}
+		FT_DEBUG("CLIENT: connected\n");
+
+	} else {
+		ct->ctrl_listenfd = socket(AF_INET, SOCK_STREAM, 0);
+		if (ct->ctrl_listenfd == -1) {
+			err = errno;
+			FT_PRINTERR("socket", err);
+		}
+		ret = setsockopt(ct->ctrl_listenfd, SOL_SOCKET, SO_REUSEADDR, &(int){ 1 }, sizeof(int));
+		if (ret == -1) {
+			err = errno;
+			FT_PRINTERR("setsockopt(SO_REUSEADDR)", err);
+			return ret;
+		}
+
+		memset(&ct->ctrl_addr, '\0', sizeof(ct->ctrl_addr));
+		ct->ctrl_addr.sin_family = AF_INET;
+		ct->ctrl_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+		ct->ctrl_addr.sin_port = htons(ct->ctrl_port);
+
+		ret = bind(ct->ctrl_listenfd, (struct sockaddr*)&ct->ctrl_addr, sizeof(ct->ctrl_addr));
+		if (ret == -1) {
+			err = errno;
+			FT_PRINTERR("bind", err);
+			return ret;
+		}
+
+		ret = listen(ct->ctrl_listenfd, 100);
+		if (ret == -1) {
+			err = errno;
+			FT_PRINTERR("listen", err);
+			return ret;
+		}
+
+		FT_DEBUG("SERVER: waiting for connection ...\n");
+		ct->ctrl_connfd = accept(ct->ctrl_listenfd, (struct sockaddr*)NULL, NULL);
+		if (ct->ctrl_connfd == -1) {
+			err = errno;
+			FT_PRINTERR("accept", err);
+			return ret;
+		}
+		FT_DEBUG("SERVER: connection acquired\n");
+	}
+
+	return 0;
+}
+
+int ft_ctrl_send(struct ct_pingpong *ct, char *buf, size_t size)
+{
+	int ret, err;
+
+	ret = write(ct->ctrl_connfd, buf, size);
+	if (ret < 0) {
+		err = errno;
+		FT_PRINTERR("write", err);
+		return ret;
+	}
+	FT_DEBUG("----> sent (%d/%ld) : \"", ret, size);
+	if (FT_ACTIVATE_DEBUG) {
+		int i;
+		for (i = 0; i < size; i++) {
+			fprintf(stderr, "%c.", buf[i]);
+		}
+		fprintf(stderr,"\"\n");
+	}
+
+	return ret;
+}
+
+int ft_ctrl_recv(struct ct_pingpong *ct, char *buf, size_t size)
+{
+	int ret, err;
+
+	ret = read(ct->ctrl_connfd, buf, size);
+	if (ret < 0) {
+		err = errno;
+		FT_PRINTERR("read", err);
+		return ret;
+	}
+	FT_DEBUG("----> received (%d/%ld) : \"", ret, size);
+	if (FT_ACTIVATE_DEBUG) {
+		int i;
+		for (i = 0; i < size; i++) {
+			fprintf(stderr, "%c.", buf[i]);
+		}
+		fprintf(stderr, "\"\n");
+	}
+
+	return ret;
+}
+
+int ft_ctrl_finish(struct ct_pingpong  *ct)
+{
+	close(ct->ctrl_connfd);
+	close(ct->ctrl_listenfd);
+
+	return 0;
+}
+
+#define FT_MSG_CHECK_PORT_OK "port ok"
+#define FT_MSG_LEN_PORT 5
+#define FT_MSG_CHECK_CNT_OK "cnt ok"
+#define FT_MSG_LEN_CNT 10
+#define FT_MSG_SYNC_Q "q"
+#define FT_MSG_SYNC_A "a"
+
+int ft_ctrl_txrx_data_port(struct ct_pingpong *ct)
+{
+	int ret;
+
+	if (ct->opts.dst_addr) {
+		memset(&ct->ctrl_buf, '\0', FT_MSG_LEN_PORT + 1);
+
+		FT_DEBUG("CLIENT: receiving port ...\n");
+		ret = ft_ctrl_recv(ct, ct->ctrl_buf, FT_MSG_LEN_PORT);
+		if (ret <= 0)
+			return EXIT_FAILURE;
+		ct->data_default_port = atoi(ct->ctrl_buf);
```

Comment:
Unless you have strict control over what is in `ct->ctrl_buf` (i'm not sure where that gets set), then I'd recommend writing a wrapper function around `strtol` and using that instead.